crates/webidl/src/util.rs

```diff
@@ -693,6 +638,149 @@ impl<'src> FirstPassRecord<'src> {
+            orig: &'a Signature<'a>,
+            args: Vec<IdlType<'a>>,
+        }
+
+        let mut actual_signatures = Vec::new();
+        'outer:
+        for signature in data.signatures.iter() {
+            let start = actual_signatures.len();
+
+            // Start off with an empty signature, this'll handle zero-argument
+            // cases and otherwise the loop below will continue to add on to this.
+            actual_signatures.push(ExpandedSig {
+                orig: signature,
+                args: Vec::with_capacity(signature.args.len()),
+            });
+
+            for (i, arg) in signature.args.iter().enumerate() {
+                let cur = actual_signatures.len();
+                // If any argument in this signature can't be converted we have
+                // to throw out the entire signature, so revert back to the
+                // beginning and then keep going.
+                let idl_type = match arg.ty.to_idl_type(self) {
+                    Some(t) => t,
+                    None => {
+                        actual_signatures.truncate(start);
+                        continue 'outer
+                    }
+                };
+
+                // The first arugment gets pushed directly in-place, but all
+                // future expanded arguments will cause new signatures to be
+                // created. If we have an optional argument then we consider the
+                // already existing signature as the "none" case and the flatten
+                // below will produce the "some" case, so we've already
+                // processed the first argument effectively.
+                let mut first = !arg.optional;
+                for idl_type in idl_type.flatten() {
+                    for j in start..cur {
+                        if first {
+                            actual_signatures[j].args.push(idl_type.clone());
+                            first = false;
+                        } else {
+                            let mut sig = actual_signatures[j].clone();
+                            sig.args.truncate(i);
+                            sig.args.push(idl_type.clone());
+                            actual_signatures.push(sig);
+                        }
+                    }
+                }
+            }
+        }
+
+        let name = match id {
+            OperationId::Constructor => "new",
+            OperationId::Operation(Some(s)) => s,
+            OperationId::Operation(None) => {
+                warn!("unsupported unnamed operation");
+                return Vec::new()
+            }
+            OperationId::IndexingGetter => "get",
+            OperationId::IndexingSetter => "set",
+            OperationId::IndexingDeleter => "delete",
+        };
+
+        let mut ret = Vec::new();
+        for signature in actual_signatures.iter() {
+            // Ignore signatures with invalid return types
+            let ret_ty = match signature.orig.ret.to_idl_type(self) {
```

Comment:
I don't think that returns are overloaded (right?) so if we did the return type filtering before the signature expansion, then we could potentially avoid doing work that will get thrown away.

Reply:
Yeah I wasn't sure about that, I'll leave a TODO for future work